### PR TITLE
feat: host groups for policy history

### PR DESCRIPTION
**Pull request by vkrizan**
Adds `host_groups` JSONB column to the PoliciesHistoryEntry.
Adds find lookups and count by host group ids for policies history entries.

Internally there are these possibilities of defining `hostGroupIds`:

* `null` -> disabled filtering, user has access to everything
* empty list -> no groups visible to the user, hence nothing should be returned
* list with a `null` (`[null]`) -> lookup for hosts that are not in a group
* list with a specific group ids -> matches only those that are in at least in one of the specified group
* combination of null and UUIDs -> the same as the two above

Should be done together with https://github.com/RedHatInsights/policies-engine/pull/546.

RHINENG-1161

### CODE REVIEW

File: src/main/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepository.java [hunk 1]
```diff
@@ -89,6 +105,40 @@ public List<PoliciesHistoryEntry> find(String orgId, UUID policyId, Pager pager)
         return query.getResultList();
     }
 
+    private static String addHostGroupsConditions(String sql, List<UUID> hostGroupIds) {
+        if (hostGroupIds == null) {
+            return sql;
+        }
+
+        List<String> conds = new ArrayList<String>();
+        conds.add("1=0");
```
[the remark this answers is not in the document]
Comment (vkrizan) [reply src/main/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepository.java:114]: Yes, this is here by design. If empty list is provided then nothing will be returned. I'll explain all cases in the next comment.

File: src/main/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepository.java [hunk 2]
```diff
@@ -89,6 +105,40 @@ public List<PoliciesHistoryEntry> find(String orgId, UUID policyId, Pager pager)
         return query.getResultList();
     }
 
+    private static String addHostGroupsConditions(String sql, List<UUID> hostGroupIds) {
+        if (hostGroupIds == null) {
```
[the remark this answers is not in the document]
Comment (vkrizan) [reply src/main/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepository.java:109]: Internally there are these possibilities:
* `null` -> disabled filtering, user has access to anything
* empty list -> no groups visible to the user, hence nothing should be returned
* list with a `null` (`[null]`) -> lookup for hosts that are not in a group
* list with a specific group ids -> matches only those that are in at least in one of the specified group
* combination of `null` and UUIDs -> the same as the two above

File: src/test/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepositoryTest.java [hunk 3]
```diff
@@ -53,7 +62,13 @@ void setUp() {
...
             helper.createPoliciesHistoryEntry(TENANT_ID_2, ORG_ID_2, POLICY_ID_2, HOST_ID_1, HOST_NAME_2, 3L),
             helper.createPoliciesHistoryEntry(TENANT_ID_2, ORG_ID_2, POLICY_ID_2, HOST_ID_2 + "-foo", HOST_NAME_2, 4L),
-            helper.createPoliciesHistoryEntry(TENANT_ID_2, ORG_ID_2, POLICY_ID_2, HOST_ID_2, "bar-" + HOST_NAME_2, 5L)
+            helper.createPoliciesHistoryEntry(TENANT_ID_2, ORG_ID_2, POLICY_ID_2, HOST_ID_2, "bar-" + HOST_NAME_2, 5L),
+
+            helper.createPoliciesHistoryEntry(TENANT_ID_3, ORG_ID_3, POLICY_ID_3, HOST_ID_1, HOST_NAME_1,1L),
+            helper.createPoliciesHistoryEntry(TENANT_ID_3, ORG_ID_3, POLICY_ID_3, HOST_ID_3, HOST_NAME_3,
+                                              List.of(GROUP_ID_1), 1L),
+            helper.createPoliciesHistoryEntry(TENANT_ID_3, ORG_ID_3, POLICY_ID_3, HOST_ID_2, HOST_NAME_2,
+                                              List.of(GROUP_ID_2), 1L)
```
[the remark this answers is not in the document]
Comment (vkrizan) [reply src/test/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepositoryTest.java:71]: I'll think about it. The default should cover it (for the most part).

File: src/test/java/com/redhat/cloud/policies/app/PoliciesHistoryTestHelper.java [hunk 4]
```diff
@@ -16,6 +22,20 @@ public class PoliciesHistoryTestHelper {
...
+        return historyEntry;
+    }
+
+    @Transactional
+    public PoliciesHistoryEntry createPoliciesHistoryEntry(String tenantId, String orgId, UUID policyId, String hostId, String hostName, List<UUID> hostGroupIds, long ctime) {
+        PoliciesHistoryEntry historyEntry = newPoliciesHistoryEntry(tenantId, orgId, policyId, hostId, hostName, ctime);
+        historyEntry.setHostGroups(hostGroupIdsToJsonArray(hostGroupIds));
+        session.persist(historyEntry);
+        return historyEntry;
+    }
```
[the remark this answers is not in the document]
Comment (vkrizan) [reply src/test/java/com/redhat/cloud/policies/app/PoliciesHistoryTestHelper.java:36]: Yeah, I guess I could have done that.

[the remark this answers is not in the document]
Comment (vkrizan) [reply src/test/java/com/redhat/cloud/policies/app/PoliciesHistoryTestHelper.java:36]: Wouldn't the two `@Transaction` methods clash? That is the reason why I've separated it, as I'm not sure what's the consequence.

File: src/test/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepositoryTest.java [hunk 5]
```diff
@@ -144,4 +159,84 @@ void testPager() {
                 .addSort("ctime", Descending).build();
         repository.find(ORG_ID_2, POLICY_ID_2, pager);
     }
+
+    @Test
+    void testHostGroupsNoResult() {
+        Pager pager = Pager.builder().build();
+        long count = repository.count(ORG_ID_3, List.of(UUID.randomUUID()), POLICY_ID_3, pager);
```
[the remark this answers is not in the document]
Comment (vkrizan) [reply src/test/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepositoryTest.java:166]: This one does not matter. The idea is to use a different, non-existent, group id.

File: src/test/java/com/redhat/cloud/policies/app/PoliciesHistoryTestHelper.java [hunk 6]
```diff
@@ -24,7 +44,14 @@ public PoliciesHistoryEntry createPoliciesHistoryEntry(String tenantId, String o
...
-        session.persist(historyEntry);
         return historyEntry;
     }
+
+    JsonArray hostGroupIdsToJsonArray(List<UUID> hostGroupIds) {
+        JsonArray ar = new JsonArray();
+        for (UUID hostGroupId : hostGroupIds) {
+            ar.add(JsonObject.of("id", hostGroupId.toString()));
+        }
+        return ar;
```
[the remark this answers is not in the document]
Comment (vkrizan) [reply src/test/java/com/redhat/cloud/policies/app/PoliciesHistoryTestHelper.java]: done

File: src/test/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepositoryTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vkrizan) [reply src/test/java/com/redhat/cloud/policies/app/model/history/PoliciesHistoryRepositoryTest.java:71]: I'm gonna keep it as is to move forward... Thank you for the suggestion.

File: src/test/java/com/redhat/cloud/policies/app/PoliciesHistoryTestHelper.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (josejulio) [reply src/test/java/com/redhat/cloud/policies/app/PoliciesHistoryTestHelper.java:36]: They won't clash, it would use the same transaction

**Review by josejulio — APPROVED** [review]
lgtm

**Pull request merged**